### PR TITLE
Porting various example commands

### PR DESCRIPTION
**Pull request by andrewlord607**
Porting different Commands Examples https://github.com/tModLoader/tModLoader/issues/1403

**ItemCommand** renamed to **ExampleItemCommand**
**NPCCommand** renamed to  **ExampleSummonCommand**
**AddTimeCommand** renamed to  **ExampleTimeCommand**
- Cleaned up the code
- Added exception handling
- Added comments

**CoinCommand** renamed to **ExampleCoinsCommand** 
and involves working with the UI so I added examples for UI

**ExampleUI** renamed to **ExampleCoinsUI**
- Refactoring code (tried to make it clearer)
- Added new comments

**DragableUIPanel** renamed to **ExampleDragableUIPanel**
**UIHoverImageButton** renamed to **ExampleUIHoverImageButton**
- Added new comments

Added **SystemUI** for handling with custom UI examples
Not sure about the name, where it should be place and how it should look

### CODE REVIEW

File: ExampleMod/Common/Systems/SystemUI.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+﻿using ExampleMod.Common.UI;
+using Microsoft.Xna.Framework;
+using System.Collections.Generic;
+using Terraria;
+using Terraria.ModLoader;
+using Terraria.UI;
+
+namespace ExampleMod.Common.Systems
+{
+	public class SystemUI: ModSystem
```
Comment (direwolf420) [inline ExampleMod/Common/Systems/SystemUI.cs:10]: missing space before `:`
```suggestion
	public class SystemUI : ModSystem
```

File: ExampleMod/Common/Systems/SystemUI.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	{
+		private UserInterface exampleCoinUserInterface;
+		internal ExampleCoinsUI exampleCoinsUI;
+
+		public override void Load() {
+			// Create custom interface wich can swap between difrent UIStates
```
Comment (direwolf420) [inline ExampleMod/Common/Systems/SystemUI.cs:16]:
```suggestion
			// Create custom interface which can swap between different UIStates
```

File: ExampleMod/Common/Systems/SystemUI.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+			exampleCoinUserInterface = new UserInterface();
+			// Creating custom UIState
+			exampleCoinsUI = new ExampleCoinsUI();
+
+			// Activate calls Initialize() on the UIState if not initialized, then calls OnActivate and then calls Activate on every child element
+			exampleCoinsUI.Activate();
+		}
+
+		public override void UpdateUI(GameTime gameTime) {
+			// Here will call .Update on custom UI and propagate it to its state and underlying elements
```
Comment (direwolf420) [inline ExampleMod/Common/Systems/SystemUI.cs:26]:
```suggestion
			// Here we call .Update on our custom UI and propagate it to its state and underlying elements
```

File: ExampleMod/Common/Systems/SystemUI.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+			if (exampleCoinUserInterface?.CurrentState != null) 
+				exampleCoinUserInterface?.Update(gameTime);
+		}
+
+		// This two methods will set the state to custom UI, causing it to show or hide(if we pass null into the method call)
+		public void ShowMyUI() {
+			exampleCoinUserInterface?.SetState(exampleCoinsUI);
+		}
+		public void HideMyUI() {
```
Comment (direwolf420) [inline ExampleMod/Common/Systems/SystemUI.cs:35]: Add a newline, and possibly move both methods to the top of the class, before the first override
```suggestion
		}
		
		public void HideMyUI() {
```

File: ExampleMod/Common/UI/ExampleDragableUIPanel.cs [hunk 5]
```diff
@@ -3,61 +3,70 @@
...
+		// Stores the offset from the top left of the UIPanel while dragging
 		private Vector2 offset;
-		public bool dragging;
+		// A flag that checks if the panel is currently being dragged
+		private bool dragging;
+
 
 		public override void MouseDown(UIMouseEvent evt) {
+			// When you override UIElements methods don't forget call base method
+			// This helps keep the basic behavior of the UIElement
```
Comment (direwolf420) [inline ExampleMod/Common/UI/ExampleDragableUIPanel.cs:23]:
```suggestion
			// When you override UIElement methods, don't forget call the base method
			// This helps to keep the basic behavior of the UIElement
```

File: ExampleMod/Common/UI/ExampleUIHoverImageButton.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+		// Tooltip text that will be shown on hover
+		internal string hoverText;
+
+		public ExampleUIHoverImageButton(Asset<Texture2D> texture, string hoverText) : base(texture) {
+			this.hoverText = hoverText;
+		}
+
+		protected override void DrawSelf(SpriteBatch spriteBatch) {
+			// When you override UIElements methods don't forget call base method
+			// This helps keep the basic behavior of the UIElement
```
Comment (direwolf420) [inline ExampleMod/Common/UI/ExampleUIHoverImageButton.cs:23]:
```suggestion
			// When you override UIElement methods, don't forget call the base method
			// This helps to keep the basic behavior of the UIElement
```

**Review by direwolf420 — COMMENTED** [review]
Some spelling and minor syntax.

File: ExampleMod/Common/Systems/SystemUI.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace ExampleMod.Common.Systems
+{
+	public class SystemUI : ModSystem
+	{
+		private UserInterface exampleCoinUserInterface;
+		internal ExampleCoinsUI exampleCoinsUI;
+        
+        // This two methods will set the state to custom UI, causing it to show or hide(if we pass null into the method call)
+		public void ShowMyUI() {
```
Comment (direwolf420) [inline ExampleMod/Common/Systems/SystemUI.cs:16]:
```suggestion
		// These two methods will set the state of our custom UI, causing it to show or hide (if we pass null into the method call)
		public void ShowMyUI() {
```

**Comment by JavidPack** [comment]
Thanks a bunch!

**Pull request merged**